vm/src/protocol/number.rs

```diff
@@ -0,0 +1,291 @@
+        add: None,
+        subtract: None,
+        multiply: None,
+        remainder: None,
+        divmod: None,
+        power: None,
+        negative: None,
+        positive: None,
+        absolute: None,
+        boolean: None,
+        invert: None,
+        lshift: None,
+        rshift: None,
+        and: None,
+        xor: None,
+        or: None,
+        int: None,
+        float: None,
+        inplace_add: None,
+        inplace_substract: None,
+        inplace_multiply: None,
+        inplace_remainder: None,
+        inplace_divmod: None,
+        inplace_power: None,
+        inplace_lshift: None,
+        inplace_rshift: None,
+        inplace_and: None,
+        inplace_xor: None,
+        inplace_or: None,
+        floor_divide: None,
+        true_divide: None,
+        inplace_floor_divide: None,
+        inplace_true_devide: None,
+        index: None,
+        matrix_multiply: None,
+        inplace_matrix_multiply: None,
+    };
+}
+
+pub struct PyNumber<'a> {
+    pub obj: &'a PyObject,
+    // some fast path do not need methods, so we do lazy initialize
+    methods: OnceCell<Cow<'static, PyNumberMethods>>,
+}
+
+impl<'a> From<&'a PyObject> for PyNumber<'a> {
+    fn from(obj: &'a PyObject) -> Self {
+        Self {
+            obj,
+            methods: OnceCell::new(),
+        }
+    }
+}
+
+impl PyNumber<'_> {
+    pub fn methods(&self, vm: &VirtualMachine) -> &PyNumberMethods {
+        &*self.methods_cow(vm)
+    }
+
+    pub fn methods_cow(&self, vm: &VirtualMachine) -> &Cow<'static, PyNumberMethods> {
+        self.methods.get_or_init(|| {
+            self.obj
+                .class()
+                .mro_find_map(|x| x.slots.as_number.load())
+                .map(|f| f(self.obj, vm))
+                .unwrap_or_else(|| Cow::Borrowed(&PyNumberMethods::NOT_IMPLEMENTED))
+        })
+    }
+
+    // PyNumber_Check
+    pub fn check(&self, vm: &VirtualMachine) -> bool {
+        let methods = self.methods(vm);
+        methods.int.is_some()
+            || methods.index.is_some()
+            || methods.float.is_some()
+            || self.obj.payload_is::<PyComplex>()
+    }
+
+    // PyIndex_Check
+    pub fn is_index(&self, vm: &VirtualMachine) -> bool {
+        self.methods(vm).index.is_some()
+    }
+
+    pub fn int(&self, vm: &VirtualMachine) -> PyResult<PyIntRef> {
+        fn try_convert(obj: &PyObject, lit: &[u8], vm: &VirtualMachine) -> PyResult<PyIntRef> {
+            let base = 10;
+            match int::bytes_to_int(lit, base) {
+                Some(i) => Ok(PyInt::from(i).into_ref(vm)),
+                None => Err(vm.new_value_error(format!(
+                    "invalid literal for int() with base {}: {}",
+                    base,
+                    obj.repr(vm)?,
+                ))),
+            }
+        }
+
+        if self.obj.class().is(PyInt::class(vm)) {
+            Ok(unsafe { self.obj.to_owned().downcast_unchecked::<PyInt>() })
```

Comment:
this part can be replaced with `payload_if_exact` to delegate unsafe operation to single spot.